src/Omega_h_pool_kokkos.cpp

```diff
@@ -0,0 +1,286 @@
+  auto [ptr, chunkIndices] = *allocationsItr;  // [begin, end)
+
+  auto [freeSetBySizeItr, freeSetByIndexItr] = insertIntoSets(chunkIndices);
+  allocations.erase(allocationsItr);
+
+  // Merge adjacent free chunks
+  if (freeSetByIndexItr != freeSetByIndex.begin()) {
+    auto prevItr = std::prev(freeSetByIndexItr);
+    auto [prevBeginIndex, prevEndIndex] = *prevItr;
+
+    if (prevEndIndex == chunkIndices.first) {
+      removeFromSets(*prevItr);
+      removeFromSets(chunkIndices);
+      freeSetByIndexItr =
+          insertIntoSets({prevBeginIndex, chunkIndices.second}).second;
+    }
+  }
+
+  if (std::next(freeSetByIndexItr) != freeSetByIndex.end()) {
+    auto nextItr = std::next(freeSetByIndexItr);
+    auto [nextBeginIndex, nextEndIndex] = *nextItr;
+    auto [beginIndex, endIndex] = *freeSetByIndexItr;
+
+    if (chunkIndices.second == nextBeginIndex) {
+      removeFromSets(*freeSetByIndexItr);
+      removeFromSets(*nextItr);
+      insertIntoSets({beginIndex, nextEndIndex});
+    }
+  }
+}
+
+auto StaticKokkosPool::getNumAllocations() const -> unsigned {
+  return allocations.size();
+}
+
+auto StaticKokkosPool::getNumFreeChunks() const -> unsigned {
+  unsigned numFreeChunks = 0;
+
+  for (const auto& [beginIndex, endIndex] : freeSetByIndex) {
+    numFreeChunks += endIndex - beginIndex;
+  }
+
+  return numFreeChunks;
+}
+
+auto StaticKokkosPool::getNumAllocatedChunks() const -> unsigned {
+  unsigned numAllocatedChunks = 0;
+
+  for (const auto& [ptr, indices] : allocations) {
+    numAllocatedChunks += indices.second - indices.first;
+  }
+
+  return numAllocatedChunks;
+}
+
+auto StaticKokkosPool::getNumChunks() const -> unsigned {
+  return numberOfChunks;
+}
+
+auto StaticKokkosPool::getNumFreeFragments() const -> unsigned {
+  return freeSetBySize.size();
+}
+
+auto StaticKokkosPool::getRequiredChunks(size_t n, size_t bytesPerChunk)
+    -> size_t {
+  return (n / bytesPerChunk) + (n % bytesPerChunk ? 1 : 0);
+}
+
+StaticKokkosPool::~StaticKokkosPool() {
+  Kokkos::kokkos_free(pool);
+}
+
+auto KokkosPool::getChunkSize() const -> size_t { return chunkSize; }
+
+auto KokkosPool::getNumFreeFragments() const -> unsigned {
+  unsigned numFreeFragments = 0;
+
+  for (const auto& pool : pools) {
+    numFreeFragments += pool.getNumFreeFragments();
+  }
+
+  return numFreeFragments;
+}
+
+KokkosPool::KokkosPool(size_t bytesPerChunks) : chunkSize(bytesPerChunks) {}
+
+auto KokkosPool::allocate(size_t n) -> void* {
+  auto current = pools.begin();
+  size_t mostAmountOfChunks = 0;
+
+  while (current != pools.end()) {
+    void* ptr = current->allocate(n);
+    if (ptr) {
+      allocations[ptr] = current;
+      return ptr;
+    }
+
+    if (current->getNumChunks() > mostAmountOfChunks) {
+      mostAmountOfChunks = current->getNumChunks();
+    }
+
+    current++;
+  }
+
+  size_t requestedChunks = StaticKokkosPool::getRequiredChunks(n, chunkSize);
+  size_t amortizedChunkSize = std::max(mostAmountOfChunks * 2, requestedChunks);
```

Comment:
We should consider making the factor of two growth a named `const` variable in a header file that 'expert' users can adjust if needed.